storage/innobase/handler/ha_innodb.cc

```diff
@@ -3860,6 +3860,17 @@ static int innodb_init_params()
 		}
 	}
 
+	if (srv_undo_tablespaces + 2 >= innobase_open_files) {
+		ib::warn() << "innodb_open_files "
+			   << innobase_open_files << " should be "
+			   << "greater than total number "
+			   << "of undo tablespace, system & temporary "
+			   << "tablespace. So changing the value of "
+			   << "innodb_open_files to "
+			   << srv_undo_tablespaces + 3;
+		innobase_open_files = srv_undo_tablespaces + 3;
```

Comment:
Can we please use `sql_print_warning()` here? Does `ib_logfile0` need to be included in the count?

For better common subexpression elimination, the condition could be written as follows:
```c++
if (srv_undo_tablespaces + 3 > innobase_open_files) {
```

Reply:
`fil_space_t::try_to_close()` tries to close only user tablespace. InnoDB doesn't create fil_space_t for redo log files.